internal/encryption/method/aesgcm/README.md

```diff
@@ -0,0 +1,54 @@
+# AES-GCM encryption method
+
+This folder contains the state encryption implementation of the AES-GCM encryption method.
+
+This is implemented following the guidance of the following document: ([NIST SP 800-38D](https://csrc.nist.gov/pubs/sp/800/38/d/final)).
+
+
+## Configuration
+
+You can configure the encryption by specifying the following method block:
+
+```hcl2
+terraform {
+  encryption {
+    method "aes_gcm" "mymethod" {
+      key        = "key here" # Pass your 16, 24, or 32 byte encryption key here.
+      aad        = ""         # Leave empty unless needed
+      nonce_size = 12         # Minimum: 1, do not change unless needed
+      tag_size   = 16         # Valid values: 12-16, do not change unless needed
+    }
+  }
+}
+```
+
+| Field              | Default | Description                                                                                                                                                                                      |
+|--------------------|---------|--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
+| `key` (*required*) |         | Your encryption key in binary format, likely referencing a key provider. Must be 16, 24, or 32 bytes long.                                                                                       |
+| `aad`              |         | Additional Authenticated Data. This data is stored along the encrypted form and authenticated. The AAD value of the encrypted form must match the configuration, otherwise the decryption fails. |
+| `nonce_size`       | 12      | Size of the Initialization Vector for AES-GCM. This field should not be changed and is provided for future compatibility.                                                                        |
+| `tag_size`         | 16      | Size of the authentication tag (12-16 bytes). This field should not be changed and is provided for future compatibility.                                                                         |
+
+## Key rotation
+
+AES-GCM keys have a limited lifetime of `2^32` blocks, equaling roughly 64 GB of data that can be encrypted before the keys should be considered compromised. Users should rotate keys well before this limit is reached.
+
+## Encryption vs. Authentication
+
+The AES-GCM implementation protects data at rest from being accessed. It does not, however, protect against malicious actors reusing old data (replay attacks) to compromise the integrity of the system. Users with the need for payload authentication should rotate their key and/or AAD frequently to ensure that old data cannot be used in this manner.

```

Comment:
Perhaps too late, but if there's no requirements for FIPS compliance, there's two alternatives to pure AES-GCM I'd suggest considering:

 - If you really want AES, choose AES-GCM-SIV, which prevents the nonce reuse problem entirely. You still have a limited number of encryptions and still have to rotate keys, but it is a much safer construction overall. Sadly NIST hasn't standardized this mode for whatever reason. This hasn't [landed in the stdlib yet](https://github.com/golang/go/issues/54364) so you're stuck using a third-party library.
 - If you're fine swapping out AES, go for XChaCha20-Poly1305. The "X" for extended allows longer nonces, so you don't have the key rotation problem at all. NIST hasn't seriously considered standardizing ChaCha20 at all to my knowledge. This is available in `x/crypto`. 

Ultimately, you still have to deal with key rotation in the event of compromise, it is just you're no longer under a risky scenario (having to count encryptions and rotate keys well ahead of hitting the limit or otherwise using time as a proxy metric).

Reply:
I think FIPS is something that's always at the back of our mind, but it is worth noting that with the way that state encryption is set up, we can have multiple different methods available for people to use.

When I was discussing this with Janos earlier, I think we discussed the fac that we want to explain some of these shortcomings with AES-GCM and help make the end users make informed choices based on their requirements. This may be another method, or more frequent key rotation.